### PR TITLE
Add support for versions of services

### PR DESCRIPTION
**Pull request by brianz**
This allows for easy subclassing of services which will allow authors to
use existing endpoints and override in a targeted fashion.  At the same
time it's trivial to add a new endpoint.  Currently there isn't a way to
remove an endpoint.

Requesting a version of a service which doesn't exists raises an error.
Clients may request a different version by simply changing the `version` argument.

### CODE REVIEW

**Comment by brianz** [comment]
@pcraciunoiu Have a look when you have some time.  Finally carved out some time to do some work to support versioning!  I've verified this is backwards compatible.

File: servant/service/base.py [hunk 1]
```diff
@@ -13,8 +13,35 @@
 from ..logger import create_logger
 
 
+class ServiceMeta(type):
+
+    def __init__(cls, name, bases, attrs):
+        action_map = ServiceMeta.__get_or_create_action_map(cls, attrs)
+
+        # Merge the action maps together for any base classes to support versioning. Ensure we walk
```
Comment (pcraciunoiu) [inline servant/service/base.py]: This seems like it could be a docstring for the class as a whole.

File: tests/test_service_class.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+def test_missing_version():
+    class NoVersion(Service):
+        name = 'service'
+        action_map = {}
+
+    with pytest.raises(Exception):
+        NoName()
+
+
+def test_missing_action_map():
```
Comment (pcraciunoiu) [inline tests/test_service_class.py]: Nice to cover your bases :)

**Comment by pcraciunoiu** [comment]
@brianz :+1: looks good. Seems quite straightforward.

Is there anything else you would want to version besides the action map? I haven't fully thought it out, but just wanted to frame the problem in case it gives you ideas of limitations of this approach.

Removing actions seems like a less urgent matter. For that I could see convenience added around deprecation warnings (e.g. v1 adds method X, v2 deprecates it, v3 removes it altogether). Maybe no convenience is necessary and just having an example pattern is enough.

File: servant/service/base.py (hunk 1 above)
Comment (brianz) [inline servant/service/base.py]: True, however this meta class may do more in the future beyond simply merging the action_maps.

**Comment by brianz** [comment]
Thanks for the feedback @pcraciunoiu 

Yeah, it's an interesting thought.  Right now `action_map` is the only thing I can think of which would need versioning.  In the case of the class methods, they handle everything for all request...so I don't think those would ever be versioned since they're service-wide.

**Pull request merged**